src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java

```diff
@@ -206,13 +231,132 @@ public void close() throws IOException
+   */
+  private void mergeWithSmoosher() throws IOException
+  {
+    // Get processed elements from the stack and write.
+    List<File> fileToProcess = new ArrayList<>(completedFiles);
+    completedFiles = Lists.newArrayList();
+    for (File file: fileToProcess)
+    {
+      add(file);
+      file.delete();
+    }
+  }
+
+  /**
+   * Returns a new SmooshedWriter which writes into temporary file and close
+   * method on returned SmooshedWriter tries to merge temporary file into
+   * original FileSmoosher object(if not open).
+   *
+   * @param name fileName
+   * @param size size of the file.
+   * @return
+   * @throws IOException
+   */
+  private SmooshedWriter delegateSmooshedWriter(final String name, final long size) throws IOException
+  {
+    final File tmpFile = new File(baseDir, name);
+    filesInProcess.add(tmpFile);
+
+    return new SmooshedWriter()
+    {
+      private int currOffset = 0;
+      private final FileOutputStream out = new FileOutputStream(tmpFile);
+      private final GatheringByteChannel channel = out.getChannel();;
+      private final Closer closer = Closer.create();
+      {
+        closer.register(out);
+        closer.register(channel);
+      }
+      @Override
+      public void close() throws IOException
+      {
+        closer.close();
+        completedFiles.add(tmpFile);
+        filesInProcess.remove(tmpFile);
+
+        if (!writerCurrentlyInUse) {
+          mergeWithSmoosher();
+        }
+      }
+      public int bytesLeft()
+      {
+        return (int) (size - currOffset);
+      }
+
+      @Override
+      public int write(ByteBuffer buffer) throws IOException
+      {
+        return addToOffset(channel.write(buffer));
+      }
+
+      @Override
+      public int write(InputStream in) throws IOException
+      {
+        return addToOffset(ByteStreams.copy(Channels.newChannel(in), channel));
+      }
+
+      @Override
+      public long write(ByteBuffer[] srcs, int offset, int length) throws IOException
+      {
+        return addToOffset(channel.write(srcs, offset, length));
+      }
+
+      @Override
+      public long write(ByteBuffer[] srcs) throws IOException
+      {
+        return addToOffset(channel.write(srcs));
+      }
+
+      public int addToOffset(long numBytesWritten)
+      {
+        if (numBytesWritten > bytesLeft()) {
+          throw new ISE("Wrote more bytes[%,d] than available[%,d]. Don't do that.", numBytesWritten, bytesLeft());
+        }
+        currOffset += numBytesWritten;
+
+        return Ints.checkedCast(numBytesWritten);
       }
+
+      @Override
+      public boolean isOpen()
+      {
+        return channel.isOpen();
+      }
+
     };
+
   }
 
   @Override
   public void close() throws IOException
   {
+    //book keeping checks on created file.
+    if (!completedFiles.isEmpty() || !filesInProcess.isEmpty())
+    {
+      for (File file: completedFiles)
+      {
+        file.delete();
+      }
+      for (File file: filesInProcess)
```

Comment:
There is an exception thrown just as soon as this for loop exits for exactly that reason. It's a bug if any files haven't yet been merged by the time the larger Smoosher gets closed (each of the individual Writers should have been closed and those should have caused everything to get merged).